File: leetcode/206.反转链表/206.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        //双指针
        // ListNode cur = head;
        // ListNode pre = null;
        // while(cur != null){
        //     ListNode tmp = cur;
        //     cur = cur.next;
        //     tmp.next = pre;
        //     pre = tmp;
        // }
        // return pre;
        //递归
        if(head == null) return null;
        if(head.next == null) return head;
        ListNode last = reverseList(head.next);
        head.next.next = head;
        head.next = null;
        return last;
    }
}

File: leetcode/206.反转链表/206.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        #双指针
        # cur = head
        # pre = = None
        # while cur:
        #     tmp = cur
        #     cur = cur.next
        #     tmp.next = pre
        #     pre = tmp
        # return pre
        # 递归
        if head == None:
            return None
        if head.next == None:
            return head
        last = self.reverseList(head.next)
        head.next.next = head
        head.next = None
        return last